sycl/doc/extensions/proposed/sycl_ext_oneapi_async_memory_alloc.asciidoc

```diff
@@ -0,0 +1,814 @@
+SYCL `handler`. None of enqueue functions return a SYCL `event` directly, as
+this extension is in line with the
+link:../experimental/sycl_ext_oneapi_enqueue_functions.asciidoc[
+  sycl_ext_oneapi_enqueue_functions] extension, so events are returned when
+calling `submit_with_event` and the `handler` overloads of these enqueue
+functions.
+
+[source,c++]
+----
+namespace ext::oneapi::experimental {
+
+void *async_malloc(queue q, usm::alloc kind, size_t size);
+
+void *async_malloc(handler h, usm::alloc kind, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+
+}  // ext::oneapi::experimental
+----
+
+[source, c++]
+----
+void *async_malloc(queue q, usm::alloc kind, size_t size);
+
+void *async_malloc(handler h, usm::alloc kind, size_t size);
+
+void *async_malloc_from_pool(queue q, size_t size, memory_pool pool);
+
+void *async_malloc_from_pool(handler h, size_t size, memory_pool pool);
+----
+
+_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
+will asynchronously allocate memory of size `size` in bytes, allocating from the
+memory pool `pool` if provided, otherwise allocation from the default memory
+pool associated with the SYCL context and device associated with `q` or `h`.
+If a memory pool `pool` is provided the allocation kind is that of the memory
+pool, otherwise if an allocation kind `kind` parameter is provided, this is the
+allocation kind which is used. If a memory pool `pool` is provided, this must be
+associated with the same SYCL context and device as `q` or the SYCL queue
+associated with `h`. Memory can either be allocated directly from the memory
+pool or allocated to the memory pool to provide enough memory in the memory pool
+for the allocation. Accessing the memory at the address of the pointer returned
+by asynchronous malloc functions before the command has completed execution is
+undefined behavior.
+
+_Returns_: A pointer to the address of a memory reservation if `size` is
+non-zero, otherwise returns `nullptr`.
+
+_Throws_: An exception with the `errc::memory_allocation` error code if the
+allocation brings the memory pool over the maximum size. This error must be
+thrown asynchronously. An exception with the `errc::invalid` error code if
+`kind` is `usm::alloc::unknown`.
+
+[source, c++]
+----
+void async_free(queue q, void *ptr);
+
+void async_free(handler h, void *ptr);
+----
+
+_Effects_: Enqueues a command to `q` or the SYCL queue associated with `h` which
+will asynchronously free the memory allocation at the address of `ptr`. Memory
+will be freed from the memory pool to be used by other asynchronous malloc
+commands which execute later, and will not free until the SYCL queue associated
+with the asynchronous allocation command has been synchronized with. Accessing
```

Comment:
This part of the sentence seems overly proscriptive:

> and will not free until the SYCL queue associated with the asynchronous allocation command has been synchronized with

I think you are trying to draw some distinction between "when the memory is freed" and "when the memory is released to the O/S".  As the first part of the sentence says, the memory is freed back to the pool as soon as the command completes.  I do not think the spec should say when the memory is released back to the O/S.